### DISCUSSION TITLE
onDone immediately called.

### DISCUSSION THREAD

**Discussion by ollyde** [enisdenjo/graphql-ws#350] [Q&A] [answered]
There doesn't seem to be any documentation on how to connect manually, without using the client side library.
I'm trying to connect to this library manually using ```web_socket_channel``` from Flutter.

When I connect using the GraphQL JS library it works, so the server is working 100%.

When I connect I get the this message from the server.

<img width="619" alt="Screenshot 2022-04-14 at 19 03 07" src="https://user-images.githubusercontent.com/6940726/163438408-6c271904-aa9a-4f3b-8fc2-8952330d8ba3.png">

I'm connecting with the following settings.

```
 _channel = WebSocketChannel.connect(
    Uri.parse(getWebStockUrl()),
    protocols: ["graphql-ws"],
  );

  final subscriptionQuery = r'''
      [Some sub query]
    '''

  _channel?.sink.add(jsonEncode({
    "type": "connection_init",
    "payload": {"Authorization": auth.jwt}
  }));
  _channel?.sink.add(jsonEncode({
    "type": "start",
    "payload": {"query": subscriptionQuery}
  }));
```

Any ideas where I'm going wrong?

**Comment by enisdenjo** [comment, marked as answer]
> There doesn't seem to be any documentation on how to connect manually, without using the client side library.

Yes there is, the [GraphQL over WebSocket Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md).

Just by looking at your example, I see you're not implementing the protocol properly. For starters, the `graphql-ws` subprotocol is used by the [deprecated `subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws).

This library uses the [`graphql-transport-ws` subprotocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#communication) (I couldn't name the subprotocol `graphql-ws` too since the deprecated library already reserved it and the two protocols are different).

**Comment by ollyde** [comment]
@enisdenjo ah we meet again. Thank you for your response, you are a life saver.

Seems we get a little further this time.

<img width="695" alt="Screenshot 2022-04-14 at 19 15 00" src="https://user-images.githubusercontent.com/6940726/163440118-eea45fb5-a267-41c3-a9e7-bbb8333d4761.png">

Not sure what I'm missing. It seems to disconnect right away. hmm

```
_channel!.stream.listen((message) {
    if (kDebugMode) {
      print("Got listen event");
      print(message);
    }
  })
    ..onData((data) {
      if (kDebugMode) {
        print("onData - WebSocket - " + data.toString().substring(0, data.toString().length > 70 ? 70 : data.toString().length) + "..");
        // print(data);
      }
      if (data is String) {
        try {
          final json = jsonDecode(data);

          if (json["type"] != null && json["type"] == "connection_ack") {
            success();
          }

          if (json?["payload"]?["data"] != null) {
            final data = json["payload"]!["data"];
            _filterData(context, data);
            return;
          }
        } catch (e) {
          //
        }
      }
    })
    ..onDone(() {
      if (kDebugMode) {
        print("onDone - WebSocket");
      }
      _reconnect(context, success);
    })
    ..onError((e) {
      if (kDebugMode) {
        print("onError - WebSocket");
        print(e);
      }
      _reconnect(context, success);
    });
```

**Reply by enisdenjo** [reply]
Please consult the [GraphQL over WebSocket Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md), there are examples at the end that can help you.

P.S. You keep logging the `context` argument of the callbacks - it is not helpful whatsoever.

**Reply by ollyde** [reply]
We're trying to migrate from ```subscriptions-transport-ws``` which seemed much simpler. :-/ what do you need?

**Reply by enisdenjo** [reply]
Sorry but I cannot write the code for you. I can, however, guarantee that if you implement the [GraphQL over WebSocket Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md), the server and the client will work coherently.

**Reply by ollyde** [reply]
@enisdenjo I understand. The server is working with the JS library. But the flutter library does not work and we are following this document, something is missing from the documentation.

```
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:uuid/uuid.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

WebSocketChannel? _channel;
bool _isReconnecting = false;

connectToWebsocket(BuildContext context, Function() success) {
  _channel?.sink.close();

  _channel = WebSocketChannel.connect(
    Uri.parse(getWebStockUrl()),
    protocols: ["graphql-transport-ws"],
  );

  final subscriptionQuery = r'''
      subscription Subscription { 
        gotChatMessage {
          messageResults {
            {ms}
          }
          room {
            {rm}
          }
        }
      }
    '''
      .replaceFirst("{ms}", messageResultFields)
      .replaceFirst("{rm}", roomFields);

  _channel?.sink.add(jsonEncode({
    "type": "connection_init",
    "payload": {"Authorization": auth.jwt}
  }));

  _channel?.sink.add(jsonEncode({
    "id": Uuid().v4(),
    "type": "start",
    "payload": {"query": subscriptionQuery}
  }));

  _channel!.stream.listen((message) {
    if (kDebugMode) {
      print("Got listen event");
      print(message);
    }
  })
    ..onData((data) {
      if (kDebugMode) {
        print(data);
      }
      if (data is String) {
        try {
          final json = jsonDecode(data);

          if (json["type"] != null && json["type"] == "connection_ack") {
            success();
          }

          if (json?["payload"]?["data"] != null) {
            final data = json["payload"]!["data"];
            _filterData(context, data);
            return;
          }
        } catch (e) {
          //
        }
      }
    })
    ..onDone(() {
      if (kDebugMode) {
        print("onDone - WebSocket");
      }
    })
    ..onError((e) {
      if (kDebugMode) {
        print("onError - WebSocket");
        print(e);
      }
    });
}

```

**Reply by enisdenjo** [reply]
> ... we are following this document, something is missing from the documentation.

Nothing is missing and you're not following the document - just by glancing at your code, I see you're still trying to subscribe using the [deprecated `subscriptions-transport-ws` Protocol GQL_START message](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md#gql_start) instead of the [`Subscribe` message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#subscribe).

Please read the [GraphQL over WebSocket Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) again, but this time with care.

**Reply by ollyde** [reply]
@enisdenjo thank you seemed to solve it. Do you want some Solana?

**Reply by enisdenjo** [reply]
Glad you got it working!

> Do you want some Solana?

Thanks a lot, but I love OSS enough to not _want_ money for my OSS work. You're free to [sponsor me (or donate) through GitHub](https://github.com/sponsors/enisdenjo) though. 😄

**Reply by ollyde** [reply]
@enisdenjo sponsored :-)
